### DISCUSSION TITLE
How to create multiple same stores without duplicating its code?

### DISCUSSION THREAD

**Discussion by igolka97** [vuejs/pinia#771] [Help and Questions] [answered]
I need to create three the same stores but with different names like this:

```
export const useCountryStore = defineStore('country', { /* options goes hire */ })
export const useRegionStore = defineStore('region', { /* options goes hire */ })
export const useCityStore = defineStore('city', { /* options goes hire */ })
```

and I want to extract options to one single place to avoid code duplication. What is the proper way to do this?

**Comment by posva** [comment, marked as answer]
create one object and copy it, just like regular js:

```js
const options = {}
defineStore('one', {...options})
defineStore('two', {...options})
```

**Reply by igolka97** [reply]
Yeah, of course I already tried it way, but TS can't find `this` object in actions while mutating state

**Reply by posva** [reply]
then create a function that returns the return of `defineStore`:


```ts
function createUseStore<Id extends string>(id: Id) {
  return defineStore(id, {
    // ...
  })
}
```

**Reply by igolka97** [reply]
Yes, this is exactly what I need. Thanks!

**Comment by Jalliuz** [comment]
I also posted the full solution on stackoverflow.com:
https://stackoverflow.com/a/78345448/5418514

It comes down to this:
````
const myStore = () => { // your store options }

export const useMyStore = (uniqueStoreName) => {
  let store = defineStore(uniqueStoreName, myStore)
  // return new store() <-- wrong
  return store()
}

export default useMyStore
````
And then in the component:
````
<template></template>
<script setup>
import useMyStore from '~/stores/myStore'
const myStore = useMyStore('a-unique-name')
</script>
````
If you do want to share the same store somewhere, you just use the same id's, otherwise a unique id

**Comment by AvWoN** [comment]
Hello @Jalliuz . Could you explain the purpose of `new` in `return new store()` instead of just returning defineStore()?

**Reply by kevinramharak** [reply]
`defineStore` returns a factory function to create a new store.
By invoking the returned factory function the code immediately creates a store and returns it.
Using the `new` here is odd, and will work because any `function` can be invoked with the `new` operator: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/new.

I would definitely recommend not doing that though, it is not how the factory function is intended to be called.

**Reply by Jalliuz** [reply]
Sorry , I just removed the "new" in "return new store()" which was not needed
